### ISSUE TITLE
BLE cannot discovery services before connect timeout. (IDFGH-7910)

### ISSUE THREAD

**Issue by Zmmfly** [espressif/esp-idf#9429]
## Environment

- Module or chip used:  
  - ESP32-WROVER-E
- IDF version
  - v4.4

## Problem Description

Cannot discovery services, when connecting to BLE 4.1 device by NimBLE library.

### Expected Behavior

Services discovered, connect success.

### Actual Behavior

Cannot discovery services, and disconnect.

**Comment by rahult-github** [comment]
Hi @Zmmfly , 

Can you please share more information 

1. Any sample application or existing IDF example that can be used to reproduce this issue ?
2. Do you have any log of your run ? Debug enabled log would be helpful.
3. Is this specific to devices with BLE 4.1 only ?
4. Is the remote 4.1 device a real word device or a proprietary solution ? 

Thanks

**Comment by Zmmfly** [comment]
Hi @rahult-github, sorry for the late reply.
1. I wrote a demo to reproduce the problem: [IDFGH-7910_demo.zip](https://yun.mxtech.ltd/s/XgNjtpXwLPgPPzp)
2. Yes, the logger.log file in the demo folder records the entire process log;
3. Yes, this problem is only found on BLE4.1 devices at present, I don't know if this is a special case, or all BLE4.1 devices will have this problem;
4. It is a real device, and you can see the Bluetooth version by observing the broadcast in nRF connect.

Thanks

**Comment by Zmmfly** [comment]
Hi @rahult-github, I don't know how to capture HCI on IDF, I used my phone to capture the HCI of the connection process, which may be of some help.
[btsnoop_hci_IDFGH7910.log](https://yun.mxtech.ltd/s/raEAzkDnkCjLgaJ)

**Comment by Zmmfly** [comment]
Great, I successfully exported [btsnoop.log](https://github.com/espressif/esp-idf/files/9199287/btsnoop_ESP32_IDFGH7910.log) on ESP32.
It looks like:
1. The first connection is disconnected by the controller before exchanging MTU;
2. The second connection is also disconnected by the controller after reading the group type request;
3. The third connection is disconnected by the controller after exchanging MTU;

First connection:
![conn1](https://user-images.githubusercontent.com/4443813/181259278-fcd98bf1-252c-43e6-97b5-e67c50551835.png)

Second connection:
![conn2](https://user-images.githubusercontent.com/4443813/181259314-133efbfb-4f4d-4210-af35-d0b9107b61d3.png)

Third connection:
![conn3](https://user-images.githubusercontent.com/4443813/181259372-1f4a28d4-5385-4365-9e80-1c62eefbb1fb.png)

**Comment by Zmmfly** [comment]
The problem has been solved, and the connection process on the device side is not standardized.

**Comment by rahult-github** [comment]
Hi @Zmmfly ,

> The problem has been solved 

Can you please help share more information about how this was solved ?.  The btsnoop log shared previously is HCI level log. So a disconnect from controller to host can be either from remote or event on the nrfapp side .

**Comment by Zmmfly** [comment]
Hi @rahult-github, according to btsnoop.log on successful connection on phone:
I found that after the connection is created (OPCode 0x2043), the MTU exchange (ATT OPCode 0x02) is not performed immediately, but the following process:
1. First initiate a request to read the remote version information (OPCode 0x041d);
2. Initiate the read group type request again (ATT OPCode 0x10);
3. After completing the service discovery, perform the MTU exchange (ATT OPCode 0x02).

According to this sequence, I modified the local NimBLE code and added an API to read the remote version information (OPCode 0x041d) request; and modified the local esp-nimble-cpp, when connecting, it is optional to perform MTU exchange or read the remote version information (OPCode 0x041d), or do nothing, and the user program completes the subsequent request.
